### PR TITLE
HPCC-22950 Add the concept of blocked time

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

### CODE REVIEW

File: common/thorhelper/thorcommon.hpp
Comment (jakesmith) [inline common/thorhelper/thorcommon.hpp]: not used.

File: thorlcr/activities/diskread/thdiskread.cpp
Comment (jakesmith) [inline thorlcr/activities/diskread/thdiskread.cpp]: my expectation was that blocked was a universal property of all activities, so would be serialized and deserialized for each, so this would be in the base class.

It's true that not all activities can block, probably about half of them can though..
_could_ have a single byte to indicate whether act. has accumulated any blocked time or not (to save other 7 bytes per act. that doesn't send.), but doubt worth it.

File: thorlcr/activities/diskread/thdiskreadslave.cpp
Comment (jakesmith) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: unfortunately, it's not as easy as this..
This will not block/will complete quickly, no matter what the data size.

aggregateStream object will be returned quickly, but calling it (nextRow) will cause blocking as it reads from the distributor inside the implementation (CAggregatingStream).
So the blocked time needs to be tracked there.

**Comment by jakesmith** [comment]
@shamser - I think you also need to take blocked time off local time (see queryLocalCycles())

**Comment by shamser** [comment]
@jakesmith @ghalliday Please can you review

File: thorlcr/activities/diskread/thdiskread.cpp
Comment (jakesmith) [inline thorlcr/activities/diskread/thdiskread.cpp]: trivial: superfluous extra newline

File: thorlcr/activities/hashdistrib/thhashdistribslave.cpp
Comment (jakesmith) [inline thorlcr/activities/hashdistrib/thhashdistribslave.cpp]: not directly related to new feature, but we did discuss changing 'totalCycles' to something else, as it's more than just that, I think called that because it used to be just a cycle_t.
Should be called something like slaveTimerStats or similar.

File: thorlcr/activities/diskread/thdiskreadslave.cpp
Comment (jakesmith) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: I think this should now be removed, there isn't really any 'blocked on others' time here. It is all local setup.
(NB: even the distributor->connect, isn't actually connecting..)

File: thorlcr/graph/thgraphmaster.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp]: queryLocalCycles() (in the slave) calculates the local time by taking total time and deducting input time.

I think it should do the same there with blocked time.
So localtime coming back already excludes blocked time.

File: common/thorhelper/thorcommon.hpp
Comment (jakesmith) [inline common/thorhelper/thorcommon.hpp]: trivial/style: normally we indent by 4 spaces here.

File: thorlcr/graph/thgraphslave.hpp
Comment (jakesmith) [inline thorlcr/graph/thgraphslave.hpp]: can be const method.

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@shamser - looks good, a few minor issues.

File: thorlcr/graph/thgraphslave.hpp
Comment (shamser) [reply thorlcr/graph/thgraphslave.hpp]: It would be nice to but serializeActivityStats is used by serializeStats and serializeStats is not const.  I can't make serializeStats const method because a derived class overrides serializeStats and that implementation updates some class variables.

File: roxie/ccd/ccdserver.cpp
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: This is new I think (didn't see in previous PR).
Not so applicable to Roxie..
Are we sure we want the albeit minor cost, of tracking what little blocked time there is in Roxie.
Perhaps should leave to a separate JIRA as none being tracked yet anyway, and we can discuss there.

File: roxie/ccd/ccdserver.cpp
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: search and replaced too much?
This isn't an ActivityTimeAccumulator type, and is actually total cycles only.
I think should not have renamed this one.

File: thorlcr/graph/thgraphmaster.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphmaster.cpp]: See other comment re. now being a pointer/Owned.

File: thorlcr/graph/thgraphslave.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphslave.cpp]: trivial/formatting: extra ;

File: thorlcr/graph/thgraphslave.cpp
Comment (jakesmith) [inline thorlcr/graph/thgraphslave.cpp]: minor: but as this should not be the norm. would be slightly better to avoid this arithmetic twice, e.g. could be:
```
    unsigned __int64 localCycles = queryTotalCycles();
    if (localCycles < inputCycles)
        return 0;
    localCycles -= inputCycles;
    const unsigned __int64 blockedCycles = queryBlockedCycles();
    if (localCycles < blockedCycles) // not sure how/if possible, but guard against
        return 0;
    return localCycles-blockedCycles;
```

File: thorlcr/graph/thgraphslave.hpp
Comment (jakesmith) [reply thorlcr/graph/thgraphslave.hpp]: >It would be nice to but serializeActivityStats is used by serializeStats and serializeStats is not const

not sure I follow, just because calling method isn't const, doesn't mean something it calls can't be...

**Comment by jakesmith** [comment]
@shamser - looks pretty close, back to you.

File: thorlcr/graph/thgraphslave.hpp
Comment (shamser) [reply thorlcr/graph/thgraphslave.hpp]: I'm not sure what I was thinking.  Probably, confused with a compiiler error message elsewhere. Fixing

File: roxie/ccd/ccdserver.cpp
Comment (shamser) [reply roxie/ccd/ccdserver.cpp]: yes, reverting this.

File: roxie/ccd/ccdserver.cpp
Comment (shamser) [reply roxie/ccd/ccdserver.cpp]: Roxie activities are calling ActivityTimeAccumulator::addStatistics().  As ::addStatistics() has been modified to update the blocked cycles, actStatistics has to be updated to include the StTimeBlocked (otherwise it fails with: assert(index < mapping.numStatistics()) failed - file: jstats.h, line 491).

File: thorlcr/graph/thgraphslave.cpp
Comment (shamser) [reply thorlcr/graph/thgraphslave.cpp]: I agree :-)

File: roxie/ccd/ccdserver.cpp
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: another over zealous s&r ? :)

File: system/jlib/jstats.cpp
Comment (jakesmith) [inline system/jlib/jstats.cpp]: pedantic: add space before }

**Comment by jakesmith** [comment]
@shamser looks good.

@richardkchapman - ready to merge after smoketest passed.

**Comment by shamser** [comment]
https://track.hpccsystems.com/browse/HPCC-22950

**Comment by shamser** [comment]
@richardkchapman Please can you merge.

File: common/thorhelper/thorcommon.hpp
Comment (ghalliday) [inline common/thorhelper/thorcommon.hpp]: If TIME_ACTIVITIES is not defined this will no compile - there needs to be a dummy implementation in the #else block.

File: roxie/ccd/ccdserver.cpp
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp]: Richard, what do you think?  Are we in danger of storing lots of 0 size blocked times.  Should ActivityTimer only add a blocked time if it is non-zero?

File: roxie/ccd/ccdserver.cpp
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: In general it would be better to have changes that rename variables (or fix whitespace issues) in separate PRs - otherwise they are in danger of hiding the significant changes.

File: thorlcr/activities/countproject/thcountprojectslave.cpp
Comment (ghalliday) [inline thorlcr/activities/countproject/thcountprojectslave.cpp]: same goes here - it is hard to review because most of the changes are simply a rename,

File: roxie/ccd/ccdserver.cpp
Comment (shamser) [reply roxie/ccd/ccdserver.cpp]: The rename was in a separate commit before Jake completed the review and asked me to squash.  I think I should able to undo the squash.

File: roxie/ccd/ccdserver.cpp
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp]: Does StCycleBlockedCycles need to be included in this list?  It is never set, but code in jlib may assume it is there.  That should possibly be changed to avoid the need to include it.

File: common/thorhelper/thorcommon.cpp
Comment (ghalliday) [inline common/thorhelper/thorcommon.cpp]: This should be conditional on blockedCycles != 0

File: common/thorhelper/thorcommon.cpp
Comment (ghalliday) [inline common/thorhelper/thorcommon.cpp]: also make this conditional

File: roxie/ccd/ccdserver.cpp
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp]: I think we should avoid adding blocked time here at all, until we have an instance that needs it.

**Review by ghalliday — COMMENTED** [review]
@shamser see other comments

**Pull request merged**
